text/0017-lmdb-vs-leveldb.md

```diff
@@ -0,0 +1,238 @@
+
+#### LevelDB
+
+The <a href="https://en.wikipedia.org/wiki/LevelDB">LevelDB Wikipedia entry</a> claims its "binary size" is 350kB. The sample Rust program that uses the <a href="https://crates.io/crates/leveldb">leveldb crate</a> in the <a href="https://github.com/mykmelez/kvbench">mykmelez/kvbench</a> repo is about 263kB larger than a control program.<sup>[3](#footnote-3)</sup>
+
+### Reliability
+
+#### LMDB
+
+According to its <a href="https://en.wikipedia.org/wiki/Lightning_Memory-Mapped_Database#Reliability">Wikipedia entry</a>, "LMDB was designed from the start to resist data loss in the face of system and application crashes. Its copy-on-write approach never overwrites currently-in-use data. [Which] means the structure on disk/storage is always valid, so application or system crashes can never leave the database in a corrupted state. In its default mode, at worst a crash can lose data from the last not-yet-committed write transaction. Even with all asynchronous modes enabled, it is only an OS catastrophic failure or hardware power-loss event rather than merely an application crash that could potentially result in any data corruption."
+
+#### LevelDB
+
+According to its <a href="https://en.wikipedia.org/wiki/LevelDB#Bugs_and_reliability">Wikipedia entry</a>, "LevelDB is widely noted for being unreliable and databases it manages are prone to corruption. Academic studies of past versions of LevelDB have found that, under some file systems, the data stored in those versions of LevelDB might become inconsistent after a system crash or power failure. LevelDB corruption is so commonplace that corruption detection has to be built into applications that use it."
+
+However, a review of the <a href="https://groups.google.com/forum/#!topic/leveldb/GXhx8YvFiig">LevelDB Issue 397 Workaround</a> discussion and some of the <a href="https://github.com/google/leveldb/issues?utf8=%E2%9C%93&q=corrupt+is%3Aissue">issues containing the string "corrupt"</a> in LevelDB's issue tracker suggests that corruption issues are taken seriously and actively investigated.
+
+### Storage Footprint
+
+In a run of the disk space "benchmark" in the [mykmelez/kvbench](https://github.com/mykmelez/kv-bench/) repo, LMDB used roughly 3.5–4x more disk space than LevelDB to store equivalent amounts of data. However, that benchmark employs a hack to measure the space taken by the LMDB and LevelDB datastores on disk, and it isn't clear that the hack is an accurate way to measure disk usage.
+
+#### LMDB
+
+LMDB doesn't compress data on disk, and it doesn't compact datastores, but it does reuse the space freed by deleted entries.
+
+#### LevelDB
+
+LevelDB compresses data on disk using <a href="https://github.com/google/snappy">Google's Snappy compression library</a>, and it compacts datastores.
+
+## References
+
+[LMDB: The Leveldb Killer?](http://banksco.de/p/lmdb-the-leveldb-killer.html)
```

Comment:
nit: LevelDB (capitalization)?

Reply:
This is actually how the blog post in question capitalizes LevelDB in its title.

(In the body of the blog post, the author uses all-lower-case "leveldb" with its first letter capitalized only at the beginning of a sentence, à la "sentence case" with the word treated as an improper noun.)